### PR TITLE
Fixes for fused batch normalizationwith FP16.

### PR DESCRIPTION
**Pull request by vcarpani**
### Summary
This PR solves compatibility problem with Tensorflow when calling operator FusedBatchNormV2 with keras.backend.floatx() set to 'float16'.
### PR Overview
As shown below Tensorflow requires that some variables (scale, offset, mean and variance) are represented in float 32bit while currently Keras is casting them to float 16bit. 
This PR cast them correctly when calling FusedBatchNormV2.

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/image_ops.cc

```c
REGISTER_OP("FusedBatchNormV2")
    .Input("x: T")
    .Input("scale: U")
    .Input("offset: U")
    .Input("mean: U")
    .Input("variance: U")
    .Output("y: T")
    .Output("batch_mean: U")
    .Output("batch_variance: U")
    .Output("reserve_space_1: U")
    .Output("reserve_space_2: U")
    .Attr("T: {half, bfloat16, float}")
    .Attr("U: {float}")
    .Attr("epsilon: float = 0.0001")
    .Attr(GetConvnetDataFormatAttrString())
    .Attr("is_training: bool = true")
    .SetShapeFn(shape_inference::FusedBatchNormShape);
``` 
- [n] This PR requires new unit tests [y/n] (make sure tests are included)
- [n] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [y] This PR is backwards compatible [y/n]
- [n] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -1926,8 +1927,8 @@ def _fused_normalize_batch_in_training(x, gamma, beta, reduction_axes,
 
     return tf.nn.fused_batch_norm(
         x,
-        gamma,
-        beta,
+        tf.cast(gamma, tf.float32),
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: You should only cast a tensor if it isn't already in the right dtype. This applies to all casts here.

Comment (vcarpani) [reply keras/backend/tensorflow_backend.py]: Changed it, however I believe that tf.cast already handles it:
```python
if isinstance(x,(ops.Tensor, _resource_variable_type)) and base_type == x.dtype:
    return x
```
https://github.com/tensorflow/tensorflow/blob/71e495448c744729c64fcc38b608a029c15828b4/tensorflow/python/ops/math_ops.py#L662

**Comment by hgaiser** [comment]
You can retrigger by closing and opening the PR.

File: keras/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -1021,6 +1021,8 @@ def moving_average_update(x, value, momentum):
     # Returns
         An operation to update the variable.
     """
+    if (value.dtype != x.dtype):
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Parens are not needed

File: keras/layers/normalization.py [hunk 3]
```diff
@@ -188,6 +188,8 @@ def normalize_inference():
             sample_size = K.prod([K.shape(inputs)[axis]
                                   for axis in reduction_axes])
             sample_size = K.cast(sample_size, dtype=K.dtype(inputs))
+            if K.backend() == 'tensorflow' and sample_size.dtype != 'float32':
```
Comment (fchollet) [inline keras/layers/normalization.py]: `sample_size.dtype != 'float32'` is always going to be true, I think, because `sample_size.dtype` is a dtype object and not a string.

Comment (vcarpani) [reply keras/layers/normalization.py]: No, I think it should work, for example if you try:
```python
a = keras.backend.arange(start=0, stop=10, step=1, dtype='int32')
a.dtype != 'int32'
```
returns False

File: keras/backend/tensorflow_backend.py (hunk 2 above)
Comment (vcarpani) [reply keras/backend/tensorflow_backend.py]: Thanks for the feedback, removed ;)

File: keras/layers/normalization.py (hunk 3 above)
Comment (hgaiser) [reply keras/layers/normalization.py]: The lines where this is handled:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/dtypes.py#L271-L272
